### PR TITLE
Add provides and uses statements to scaffold module-info

### PR DESCRIPTION
**Pull request by mocobeta**
Follow-up of https://github.com/dweiss/lucene/pull/4

```
lucene $ ./gradlew -p lucene/core/ scaffoldModuleDescriptor

> Task :lucene:core:scaffoldModuleDescriptor
Output to: /mnt/hdd/repo/lucene/lucene/core/build/tmp/scaffoldModuleDescriptor/module-info.java
NOTE: The generated module descriptor is not complete and won't work as is.

lucene $ less /mnt/hdd/repo/lucene/lucene/core/build/tmp/scaffoldModuleDescriptor/module-info.java
module org.apache.lucene.core {
  exports org.apache.lucene.analysis;
  exports org.apache.lucene.analysis.standard;
  exports org.apache.lucene.analysis.tokenattributes;
  exports org.apache.lucene.codecs;
  exports org.apache.lucene.codecs.compressing;
  exports org.apache.lucene.codecs.lucene90;
  exports org.apache.lucene.codecs.lucene90.blocktree;
  exports org.apache.lucene.codecs.lucene90.compressing;
  exports org.apache.lucene.codecs.perfield;
  exports org.apache.lucene.document;
  exports org.apache.lucene.geo;
  exports org.apache.lucene.index;
  exports org.apache.lucene.search;
  exports org.apache.lucene.search.comparators;
  exports org.apache.lucene.search.similarities;
  exports org.apache.lucene.store;
  exports org.apache.lucene.util;
  exports org.apache.lucene.util.automaton;
  exports org.apache.lucene.util.bkd;
  exports org.apache.lucene.util.compress;
  exports org.apache.lucene.util.fst;
  exports org.apache.lucene.util.graph;
  exports org.apache.lucene.util.hnsw;
  exports org.apache.lucene.util.hppc;
  exports org.apache.lucene.util.mutable;
  exports org.apache.lucene.util.packed;

  provides org.apache.lucene.codecs.DocValuesFormat with
      org.apache.lucene.codecs.lucene90.Lucene90DocValuesFormat;
  provides org.apache.lucene.codecs.PostingsFormat with
      org.apache.lucene.codecs.lucene90.Lucene90PostingsFormat;
  provides org.apache.lucene.analysis.TokenizerFactory with
      org.apache.lucene.analysis.standard.StandardTokenizerFactory;
  provides org.apache.lucene.index.SortFieldProvider with
      org.apache.lucene.search.SortField.Provider,
      org.apache.lucene.search.SortedNumericSortField.Provider,
      org.apache.lucene.search.SortedSetSortField.Provider;
  provides org.apache.lucene.codecs.Codec with
      org.apache.lucene.codecs.lucene90.Lucene90Codec;
  provides org.apache.lucene.codecs.KnnVectorsFormat with
      org.apache.lucene.codecs.lucene90.Lucene90HnswVectorsFormat;

  uses org.apache.lucene.codecs.DocValuesFormat;
  uses org.apache.lucene.codecs.PostingsFormat;
  uses org.apache.lucene.analysis.TokenizerFactory;
  uses org.apache.lucene.index.SortFieldProvider;
  uses org.apache.lucene.codecs.Codec;
  uses org.apache.lucene.codecs.KnnVectorsFormat;
}
```

Note: Services that have no implementations won't be shown.

### CODE REVIEW

**Comment by dweiss** [comment]
Hmm... How many Lucene JARs are there? Wouldn't it be easier to just manually port them one by one? With all the automation, you'll still have to add the requires-something-dependency statements anyway so there is going to be manual work involved for each and every module. I'm not against automatization but it seems like moving at a slower pace and modifying one module at a time is going to be needed anyway to ensure everything works.

**Comment by mocobeta** [comment]
I don't intend to automate everything, this is basically for analysis modules that have many services that should be copied from META-INF/services (if I miss copying them, they are silently ignored and there would not be any chance to notice that).
I'm going to finish tweaking the scaffolding staff with the change...

**Comment by mocobeta** [comment]
But if more proper automation and/or validation will come, this wouldn't be needed.

**Comment by dweiss** [comment]
ok. should I merge it in then?

**Comment by mocobeta** [comment]
For the time being, it would be useful for me. I think we can discard the task at some point (when we complete the PR).

**Comment by mocobeta** [comment]
These are temporal use anyway; I surely can keep the code on my local branch, so this can be left here if you don't like it.

**Comment by dweiss** [comment]
I like it. I just don't think it should be left in for the final merge: people will wonder what the heck this thing is for. :)

**Comment by uschindler** [comment]
Hi @mocobeta: Please remove all "uses" statements in all modules, except in "org.apache.lucene.core" module. I have seen that you added it to the scaffold tool.

"uses" means that the services are consumed by ServiceLoader. The modules who provide them are not using them. Lucene reads the services only in "core" by calling ServiceLoader. The code for that is all in core: `NamedSPILoader` (used by codec components) and in `AnalysisSPILoader`. We don't load services anywhere else!

So Lucene core must list all services as "uses" only. The complicated thing is: core has both (it consumes as only one and provides, but that's unrelated).

File: gradle/java/modules.gradle [hunk 1]
```diff
@@ -71,20 +71,53 @@ allprojects {
...
+            def service = entry.key
+            def providers = "      " + entry.value.join(",\n      ")
+            writer.write("  provides ")
+            writer.write(service)
+            writer.write(" with\n")
+            writer.write(providers)
+            writer.write(";\n")
+          }}
+          writer.write("\n")
+          // write uses statements
```
Comment (uschindler) [inline gradle/java/modules.gradle:117]: This must go away for all modules except core. We only consume ALL services from all modules in lucene's core.

**Comment by mocobeta** [comment]
@uschindler ah right, sure. I will remove all uses statements.

**Pull request merged**